From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Meow <mdev73677@gmail.com>
Date: Mon, 3 Aug 2026 22:01:07 +0200
Subject: [PATCH] 9510: decompiler: allow non-first output register as return
 value in fallback path

Allow a non-first output register to be the return value
---
 .../Decompiler/certification.manifest         |  1 +
 .../Decompiler/src/decompile/cpp/fspec.cc     |  2 +-
 .../fallback_return_nonfirst_register.xml     | 30 +++++++++++++++++++
 3 files changed, 32 insertions(+), 1 deletion(-)
 create mode 100644 Ghidra/Features/Decompiler/src/decompile/datatests/fallback_return_nonfirst_register.xml

diff --git a/Ghidra/Features/Decompiler/certification.manifest b/Ghidra/Features/Decompiler/certification.manifest
index 7636244cf1..58b0880cf7 100644
--- a/Ghidra/Features/Decompiler/certification.manifest
+++ b/Ghidra/Features/Decompiler/certification.manifest
@@ -42,6 +42,7 @@ src/decompile/datatests/doublemove.xml||GHIDRA||||END|
 src/decompile/datatests/dupptr.xml||GHIDRA||||END|
 src/decompile/datatests/elseif.xml||GHIDRA||||END|
 src/decompile/datatests/enum.xml||GHIDRA||||END|
+src/decompile/datatests/fallback_return_nonfirst_register.xml||GHIDRA||||END|
 src/decompile/datatests/floatcast.xml||GHIDRA||||END|
 src/decompile/datatests/floatconv.xml||GHIDRA||||END|
 src/decompile/datatests/floatprint.xml||GHIDRA||||END|
diff --git a/Ghidra/Features/Decompiler/src/decompile/cpp/fspec.cc b/Ghidra/Features/Decompiler/src/decompile/cpp/fspec.cc
index 69ac19b5fd..64bb95c122 100644
--- a/Ghidra/Features/Decompiler/src/decompile/cpp/fspec.cc
+++ b/Ghidra/Features/Decompiler/src/decompile/cpp/fspec.cc
@@ -1846,7 +1846,7 @@ void ParamListStandardOut::fillinMap(ParamActive *active) const
       return;
     }
   }
-  fillinMapFallback(active, true);
+  fillinMapFallback(active, false);
 }
 
 bool ParamListStandardOut::possibleParam(const Address &loc,int4 size) const
diff --git a/Ghidra/Features/Decompiler/src/decompile/datatests/fallback_return_nonfirst_register.xml b/Ghidra/Features/Decompiler/src/decompile/datatests/fallback_return_nonfirst_register.xml
new file mode 100644
index 0000000000..05f4bd294d
--- /dev/null
+++ b/Ghidra/Features/Decompiler/src/decompile/datatests/fallback_return_nonfirst_register.xml
@@ -0,0 +1,30 @@
+<decompilertest>
+  <!--The cspec.xml documentation states:
+    "In the inverse case, the decompiler examines all (possible) output varnodes that have
+    a corresponding pentry tag in the resource list. The varnode whose corresponding
+    tag occurs the earliest in the list becomes the formal return value for the function.
+    If an output varnode matches no pentry, then it is rejected as a formal return value."
+
+    Bytes are `clang -O2 -msse` output.
+    // https://godbolt.org/z/xW5hEz91T
+
+    The decompiler sees only XMM1 written.
+    The engine should not reject XMM1 as a non-first in class float (XMM0 is first) and emits
+    `void test_function(void)`.
+
+-->
+  <binaryimage arch="x86:LE:64:default:gcc">
+    <bytechunk space="ram" offset="0x100000" readonly="true">
+      f20f104c2408c3
+</bytechunk>
+    <symbol space="ram" offset="0x100000" name="test_function" />
+  </binaryimage>
+  <script>
+  <com>lo fu test_function</com>
+  <com>decompile</com>
+  <com>print C</com>
+  <com>quit</com>
+</script>
+  <stringmatch name="matching pentry becomes return" min="1" max="1">return in_stack</stringmatch>
+  <stringmatch name="test_function is not void" min="0" max="0">void test_function</stringmatch>
+</decompilertest>
-- 
2.45.1

